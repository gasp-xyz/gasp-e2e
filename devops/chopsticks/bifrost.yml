endpoint: wss://bifrost-rpc.dwellir.com
mock-signature-host: true
db: ./bifrost.db.sqlite
port: 9948

import-storage:
  System:
    Account:
      -
        -
          - 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
        - data:
            free: 1000000000000000 # 1K BNC
  Tokens:
    Accounts:
      -
        -
          - 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
          - vtoken: KSM
        - free: 1000000000000000 # 1K vKSM
      -
        -
          - 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
          - vtoken: BNC
        - free: 1000000000000000 # 1K vBNC
      -
        -
          - 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
          - vtoken: MOVR
        - free: '1000000000000000000000' # 1K vMOVR
      -
        -
          - 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
          - vstoken: KSM
        - free: 1000000000000000 # 1K vsKSM
      -
        -
          - 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
          - token: ZLK
        - free: '1000000000000000000000' # 1K ZLK
  Council:
    Members: [5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY]