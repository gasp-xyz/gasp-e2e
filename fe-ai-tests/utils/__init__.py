# This file makes the utils directory a Python package
# This allows imports like: from utils.common_test_utils import x
