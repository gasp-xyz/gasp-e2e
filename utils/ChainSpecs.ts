import { BN_TEN, BN_ZERO } from "@mangata-finance/sdk";
import { BN, BN_HUNDRED, BN_THOUSAND } from "@polkadot/util";
import { ApiPromise } from "@polkadot/api";

export const TRANSFER_INSTRUCTIONS = 4;
export const WEIGHT_IN_SECONDS = new BN(1_000_000_000_000);

export interface ChainSpec {
  unitCostWeight: number;
  parachain: number;
  assets: Map<AssetSpec, AssetMeta>;
  foreign: Map<AssetSpec, AssetMeta>;
}

export interface AssetSpec {
  symbol: string;
  decimals: number;
  location: object;
  unit: BN;
}

interface AssetMeta {
  fps: BN;
  ed: BN;
  location: object;
}

export enum ChainId {
  Mg,
  Tur,
  Bifrost,
  Statemine,
}

export class AssetId {
  static Tur: AssetSpec = {
    symbol: "TUR",
    decimals: 10,
    location: { parents: 1, interior: { X1: { Parachain: 2114 } } },
    unit: BN_TEN.pow(new BN(10)),
  };
  static TurV3: AssetSpec = {
    symbol: "TUR",
    decimals: 10,
    location: { parents: 1, interior: { X1: { Parachain: 2114 } } },
    unit: BN_TEN.pow(new BN(10)),
  };

  static Mgx: AssetSpec = {
    symbol: "MGX",
    decimals: 18,
    location: {
      parents: 1,
      interior: { X2: [{ Parachain: 2110 }, { GeneralKey: "0x00000000" }] },
    },
    unit: BN_TEN.pow(new BN(18)),
  };
  static async getLocationFromChain(symbol: string, api: ApiPromise) {
    const entries = await api.query.assetRegistry.metadata.entries();
    const asset = entries.filter((entry) => {
      const elem = entry[1];
      const name = api.createType(
        "Vec<u8>",
        JSON.parse(JSON.stringify(elem)).symbol,
      );
      return name.toHuman() === symbol;
    })[0];
    return JSON.parse(JSON.stringify(asset[1].toHuman())).location;
  }
  static Bnc: AssetSpec = {
    symbol: "BNC",
    decimals: 12,
    location: {
      parents: 1,
      interior: { X2: [{ Parachain: 2001 }, { GeneralKey: "0x0001" }] },
    },
    unit: BN_TEN.pow(new BN(12)),
  };
  static ImbueBncV3: AssetSpec = {
    symbol: "BNC",
    decimals: 12,
    location: {
      parents: 0,
      interior: {
        X1: [
          {
            GeneralKey: {
              length: 2,
              data: "0x0001000000000000000000000000000000000000000000000000000000000000",
            },
          },
        ],
      },
    },
    unit: BN_TEN.pow(new BN(12)),
  };
  static BncV3: AssetSpec = {
    symbol: "BNC",
    decimals: 12,
    location: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: 2001 },
          {
            GeneralKey: {
              length: 2,
              data: "0x0001000000000000000000000000000000000000000000000000000000000000",
            },
          },
        ],
      },
    },
    unit: BN_TEN.pow(new BN(12)),
  };

  static USDt: AssetSpec = {
    symbol: "USDt",
    decimals: 6,
    location: {
      parents: 1,
      interior: {
        X3: [
          { Parachain: 1000 },
          { PalletInstance: 50 },
          { GeneralIndex: 1984 },
        ],
      },
    },
    unit: BN_TEN.pow(new BN(6)),
  };
}

export const ChainSpecs = new Map<ChainId, ChainSpec>([
  [
    ChainId.Tur,
    {
      unitCostWeight: 1_000_000_000,
      parachain: 2114,
      assets: new Map([
        [
          AssetId.Tur,
          {
            fps: new BN(416_000_000_000),
            ed: AssetId.Tur.unit.mul(BN_TEN),
            location: AssetId.Tur.location,
          },
        ],
      ]),
      foreign: new Map(),
    },
  ],
  [
    ChainId.Mg,
    {
      unitCostWeight: 150_000_000,
      parachain: 2110,
      assets: new Map(),
      foreign: new Map([
        [
          AssetId.Tur,
          {
            fps: new BN(537_600_000_000),
            ed: BN_ZERO,
            location: AssetId.Tur.location,
          },
        ],
        [
          AssetId.TurV3,
          {
            fps: new BN(537_600_000_000),
            ed: BN_ZERO,
            location: AssetId.TurV3.location,
          },
        ],
        [
          AssetId.Bnc,
          {
            fps: new BN(43008000000000),
            ed: BN_ZERO,
            location: AssetId.Bnc.location,
          },
        ],
        [
          AssetId.ImbueBncV3,
          {
            fps: new BN(43008000000000),
            ed: BN_ZERO,
            location: AssetId.BncV3.location,
          },
        ],
        [
          AssetId.BncV3,
          {
            fps: new BN(43008000000000),
            ed: BN_ZERO,
            location: AssetId.BncV3.location,
          },
        ],
        [
          AssetId.USDt,
          {
            fps: new BN(13440000),
            ed: BN_ZERO,
            location: AssetId.USDt.location,
          },
        ],
      ]),
    },
  ],
  [
    ChainId.Bifrost,
    {
      unitCostWeight: 200_000_000,
      parachain: 2001,
      assets: new Map([
        [
          AssetId.Bnc,
          {
            fps: new BN(9360000000000),
            ed: AssetId.Bnc.unit.div(BN_HUNDRED),
            location: {
              parents: 0,
              interior: { X1: { GeneralKey: "0x0001" } },
            },
          },
        ],
        [
          AssetId.BncV3,
          {
            fps: new BN(9360000000000),
            ed: AssetId.BncV3.unit.div(BN_HUNDRED),
            location: {
              parents: 0,
              interior: {
                X1: {
                  GeneralKey: {
                    length: 2,
                    data: "0x0001000000000000000000000000000000000000000000000000000000000000",
                  },
                },
              },
            },
          },
        ],
        [
          AssetId.ImbueBncV3,
          {
            fps: new BN(9360000000000),
            ed: AssetId.ImbueBncV3.unit.div(BN_HUNDRED),
            location: {
              parents: 0,
              interior: {
                X1: {
                  GeneralKey: {
                    length: 2,
                    data: "0x0001000000000000000000000000000000000000000000000000000000000000",
                  },
                },
              },
            },
          },
        ],
      ]),
      foreign: new Map(),
    },
  ],
  [
    ChainId.Statemine,
    {
      unitCostWeight: 1_000_000_000,
      parachain: 1000,
      assets: new Map([
        [
          AssetId.USDt,
          {
            fps: new BN(1),
            ed: AssetId.USDt.unit.div(BN_THOUSAND),
            location: {
              parents: 0,
              interior: {
                X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }],
              },
            },
          },
        ],
      ]),
      foreign: new Map(),
    },
  ],
]);
