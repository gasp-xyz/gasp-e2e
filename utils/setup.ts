import { ApiPromise, Keyring } from "@polkadot/api";
import { getEnvironmentRequiredVars } from "./utils";
import { User } from "./User";
import "@mangata-finance/types";
import { getApi, initApi } from "./api";
import { Sudo } from "./sudo";
import { Assets } from "./Assets";
import { Xyk } from "./xyk";
import { SudoDB } from "./SudoDB";
import { Codec } from "@polkadot/types-codec/types";
import { signTx } from "@mangata-finance/sdk";
import { SubmittableExtrinsic } from "@polkadot/api/types";
// API
export let api: ApiPromise;

// Users
export let keyring: Keyring;
export let sudo: User;
export let alice: User;
export let eve: User;

export type Extrinsic = SubmittableExtrinsic<"promise">;

export type CodecOrArray = Codec | Codec[];

const processCodecOrArray = (codec: CodecOrArray, fn: (c: Codec) => any) =>
  Array.isArray(codec) ? codec.map(fn) : fn(codec);

export const toHuman = (codec: CodecOrArray) =>
  processCodecOrArray(codec, (c) => c?.toHuman?.() ?? c);
export const toJson = (codec: CodecOrArray) =>
  processCodecOrArray(codec, (c) => c?.toJSON?.() ?? c);
export const toHex = (codec: CodecOrArray) =>
  processCodecOrArray(codec, (c) => c?.toHex?.() ?? c);

export const setupApi = async () => {
  if (!api || (api && !api.isConnected)) {
    await initApi();
    api = getApi();
  }
};
export function isBackendTest() {
  const groupPrefix = "--group=";
  const isThereAPath = process.argv.find((arg) => arg.includes("test/"));
  const isAGroupRun = process.argv.find((arg) => arg.includes(groupPrefix));
  if (
    isThereAPath &&
    isThereAPath.length > 0 &&
    isThereAPath.toLowerCase().includes("ui")
  ) {
    return false;
  }
  return !(
    isAGroupRun &&
    isAGroupRun.length > 0 &&
    isAGroupRun.toLowerCase().includes("ui")
  );
}
export const setupUsers = () => {
  keyring = new Keyring({ type: "sr25519" });
  const { sudo: sudoUserName } = getEnvironmentRequiredVars();
  sudo = new User(keyring, sudoUserName);
  alice = new User(keyring, "//Alice");
  eve = new User(keyring, "//Eve");
  const testUser1 = new User(keyring);
  const testUser2 = new User(keyring);
  const testUser3 = new User(keyring);
  const testUser4 = new User(keyring);
  const testUser5 = new User(keyring);
  const testUser6 = new User(keyring);

  keyring.addPair(sudo.keyRingPair);
  keyring.addPair(alice.keyRingPair);
  keyring.addPair(eve.keyRingPair);
  keyring.addPair(testUser1.keyRingPair);
  keyring.addPair(testUser2.keyRingPair);
  keyring.addPair(testUser3.keyRingPair);
  keyring.addPair(testUser4.keyRingPair);
  keyring.addPair(testUser5.keyRingPair);
  keyring.addPair(testUser6.keyRingPair);

  return [testUser1, testUser2, testUser3, testUser4, testUser5, testUser6];
};

export const devTestingPairs = (ss58Format?: number) => {
  const keyring = new Keyring({ type: "sr25519", ss58Format });
  const alice = keyring.addFromUri("//Alice");
  const bob = keyring.addFromUri("//Bob");
  const charlie = keyring.addFromUri("//Charlie");
  const dave = keyring.addFromUri("//Dave");
  const eve = keyring.addFromUri("//Eve");
  return {
    alice,
    bob,
    charlie,
    dave,
    eve,
    keyring,
  };
};

export async function setup5PoolsChained(users: User[]) {
  const [testUser1, testUser2, testUser3, testUser4] = await setupUsers();
  users = [testUser1, testUser2, testUser3, testUser4];
  const keyring = new Keyring({ type: "sr25519" });
  const sudo = new User(keyring, getEnvironmentRequiredVars().sudo);
  const tokenIds = await SudoDB.getInstance().getTokenIds(5);
  const mints = [
    Assets.mintToken(tokenIds[0], sudo),
    Assets.mintToken(tokenIds[1], sudo),
    Assets.mintToken(tokenIds[2], sudo),
    Assets.mintToken(tokenIds[3], sudo),
    Assets.mintToken(tokenIds[4], sudo),
  ];

  const poolCreationExtrinsics: Extrinsic[] = [];
  tokenIds.forEach((_, index, tokens) => {
    poolCreationExtrinsics.push(
      Xyk.createPool(
        tokenIds[index],
        Assets.DEFAULT_AMOUNT.divn(2),
        tokenIds[index + (1 % tokens.length)],
        Assets.DEFAULT_AMOUNT.divn(2),
      ),
    );
  });
  await Sudo.batchAsSudoFinalized(
    ...mints,
    Assets.mintNative(testUser1),
    Assets.mintNative(testUser2),
    Assets.mintNative(testUser3),
    Assets.mintNative(testUser4),
    Assets.mintToken(tokenIds[0], testUser1),
    Assets.mintToken(tokenIds[0], testUser2),
    Assets.mintToken(tokenIds[0], testUser3),
    Assets.mintToken(tokenIds[0], testUser4),
    Assets.mintToken(tokenIds[tokenIds.length - 1], testUser1),
    Assets.mintToken(tokenIds[tokenIds.length - 1], testUser2),
    Assets.mintToken(tokenIds[tokenIds.length - 1], testUser3),
    Assets.mintToken(tokenIds[tokenIds.length - 1], testUser4),
    ...poolCreationExtrinsics,
  );
  return { users, tokenIds };
}
export async function setupAPoolForUsers(users: User[]) {
  const [testUser1, testUser2, testUser3, testUser4] = await setupUsers();
  users = [testUser1, testUser2, testUser3, testUser4];
  const keyring = new Keyring({ type: "sr25519" });
  const sudo = new User(keyring, getEnvironmentRequiredVars().sudo);

  const tokenIds = await SudoDB.getInstance().getTokenIds(2);

  const poolCreationExtrinsics: Extrinsic[] = [];
  poolCreationExtrinsics.push(
    Xyk.createPool(
      tokenIds[0],
      Assets.DEFAULT_AMOUNT.divn(2),
      tokenIds[1],
      Assets.DEFAULT_AMOUNT.divn(2),
    ),
  );

  await Sudo.batchAsSudoFinalized(
    Assets.mintToken(tokenIds[0], sudo),
    Assets.mintToken(tokenIds[1], sudo),
    Assets.mintNative(testUser1),
    Assets.mintNative(testUser2),
    Assets.mintNative(testUser3),
    Assets.mintNative(testUser4),
    Assets.mintToken(tokenIds[0], testUser1),
    Assets.mintToken(tokenIds[0], testUser2),
    Assets.mintToken(tokenIds[0], testUser3),
    Assets.mintToken(tokenIds[0], testUser4),
    ...poolCreationExtrinsics,
  );
  return { users, tokenIds };
}
export const setupGasLess = async (force = false) => {
  keyring = new Keyring({ type: "sr25519" });
  const { sudo: sudoUserName } = getEnvironmentRequiredVars();
  sudo = new User(keyring, sudoUserName);
  alice = new User(keyring, "//Alice");
  await setupApi();
  const feeLockConfig = JSON.parse(
    JSON.stringify(await api?.query.feeLock.feeLockMetadata()),
  );
  // only create if empty.
  if (feeLockConfig === null || feeLockConfig.periodLength === null || force) {
    const extrinsic = api!.tx.feeLock
      .updateFeeLockMetadata(
        10,
        "50000000000000000000",
        "1000000000000000000000",
        [[1, false]],
      )
      .toString();
    await signTx(api!, api!.tx.sudo.sudo(extrinsic), sudo.keyRingPair, {
      nonce: await SudoDB.getInstance().getSudoNonce(sudo.keyRingPair.address),
    });
  }
};
