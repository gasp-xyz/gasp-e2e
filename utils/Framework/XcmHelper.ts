import {
  Blockchain,
  BuildBlockMode,
  setupWithServer,
} from "@acala-network/chopsticks";
import { StorageValues } from "@acala-network/chopsticks/lib/utils/set-storage";
import { ApiPromise, WsProvider } from "@polkadot/api";
import { HexString } from "@polkadot/util/types";
import { getPort } from "get-port-please";
import * as fs from "fs";
import { bufferToU8a, u8aToHex } from "@polkadot/util";
import { Assets } from "../Assets";
import { alice } from "../setup";
import { Sudo } from "../sudo";

export type SetupOption = {
  endpoint: string;
  blockNumber?: number;
  blockHash?: HexString;
  wasmOverride?: string;
  db?: string;
  types?: Record<string, any>;
  localPort?: number;
  buildBlockMode?: BuildBlockMode;
};

export type DevApi = {
  newBlock: (param?: { count?: number; to?: number }) => Promise<string>;
  setStorage: (values: StorageValues, blockHash?: string) => Promise<void>;
  timeTravel: (date: string | number) => Promise<number>;
  setHead: (hashOrNumber: string | number) => Promise<void>;
};

export type ApiContext = {
  uri: string;
  chain: Blockchain;
  ws: WsProvider;
  api: ApiPromise;
  dev: DevApi;
  teardown: () => Promise<void>;
};

export const setupContext = async ({
  endpoint,
  blockNumber,
  blockHash,
  wasmOverride,
  db,
  types,
  localPort,
  buildBlockMode,
}: SetupOption): Promise<ApiContext> => {
  // random port
  const port = localPort ? localPort : await getPort();
  const config = {
    endpoint,
    port,
    block: blockNumber || blockHash,
    mockSignatureHost: true,
    "build-block-mode": buildBlockMode ? buildBlockMode : BuildBlockMode.Manual,
    db,
    "wasm-override": wasmOverride,
    "registered-types": { types: types },
    "runtime-log-level": 5,
    runtimeLogLevel: 5,
  };
  const { chain, listenPort, close } = await setupWithServer(config);
  const uri = `ws://localhost:${listenPort}`;
  const ws = new WsProvider(uri);
  const api = await ApiPromise.create({
    provider: ws,
    types: types,
  });

  await api.isReady;

  return {
    uri,
    chain,
    ws,
    api,
    dev: {
      newBlock: (param?: { count?: number; to?: number }): Promise<string> => {
        return ws.send("dev_newBlock", [param]);
      },
      setStorage: (values: StorageValues, blockHash?: string) => {
        return ws.send("dev_setStorage", [values, blockHash]);
      },
      timeTravel: (date: string | number) => {
        return ws.send<number>("dev_timeTravel", [date]);
      },
      setHead: (hashOrNumber: string | number) => {
        return ws.send("dev_setHead", [hashOrNumber]);
      },
    },
    async teardown() {
      await api.disconnect();
      await close();
    },
  };
};
export async function upgradeMangata(mangata: ApiContext) {
  const path = `test/xcm/_releasesUT/0.30.0/mangata_kusama_runtime-0.30.0.RC.compact.compressed.wasm`;
  const wasmContent = fs.readFileSync(path);
  const hexHash = mangata.api!.registry.hash(bufferToU8a(wasmContent)).toHex();
  await Sudo.batchAsSudoFinalized(Assets.mintNative(alice));
  await Sudo.asSudoFinalized(
    Sudo.sudo(
      //@ts-ignore
      mangata.api!.tx.parachainSystem.authorizeUpgrade(hexHash)
    )
  );
  const wasmParam = Uint8Array.from(wasmContent);
  const hex = u8aToHex(wasmParam);
  const param = hex.toString();
  await mangata.api.tx.sudo
    .sudo(mangata.api.tx.parachainSystem.enactAuthorizedUpgrade(param))
    .signAndSend(alice.keyRingPair);

  await mangata.dev.newBlock();
  await mangata.dev.newBlock();
}
