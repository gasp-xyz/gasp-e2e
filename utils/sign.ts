// wrap SDK signTx to allow verbose logs
import { MangataGenericEvent, signTx } from "@mangata-finance/sdk";
import { ApiPromise } from "@polkadot/api";
import { Codec } from "@polkadot/types/types";
import { BN } from "@polkadot/util";
import _ from "lodash";
import { ExtrinsicResult } from "./eventListeners";
import { logEvent, testLog } from "./Logger";
import { api, Extrinsic } from "./setup";
import { getEventResultFromMangataTx } from "./txHandler";
import { User } from "./User";

export const signSendFinalized = async (
  tx: Extrinsic,
  user: User,
  nonce: BN | undefined = undefined,
): Promise<MangataGenericEvent[]> => {
  return signTx(api, tx, user.keyRingPair, {
    nonce: nonce,
    statusCallback: ({ events = [], status }) => {
      testLog.getLog().info(status);
      events.forEach((e) => logEvent("mangata", e));
    },
  })
    .catch((reason) => {
      testLog.getLog().error(reason.data || reason);
      throw reason;
    })
    .then((result) => {
      const event = getEventResultFromMangataTx(result);
      if (event.state === ExtrinsicResult.ExtrinsicFailed) {
        throw event;
      }
      return result;
    });
};

// used for APIs other than mangata
export const signSendSuccess = async (
  api: ApiPromise,
  tx: Extrinsic,
  user: User,
): Promise<void> => {
  return new Promise(async (resolve, reject) => {
    try {
      const unsub = await tx.signAndSend(
        user.keyRingPair,
        ({ events, status, dispatchError }) => {
          testLog
            .getLog()
            .info(`â†’ events on ${api.runtimeChain} for ${status}`);

          events.forEach((e) => logEvent(api.runtimeChain, e));

          if (!_.isNil(dispatchError)) {
            if (dispatchError.isModule) {
              const metaError = api.registry.findMetaError(
                dispatchError.asModule,
              );
              const { name, section } = metaError;
              reject(new Error(`${section}.${name}`));
              return;
            } else {
              reject(new Error(dispatchError.toString()));
              return;
            }
          }

          const event = _.find(events, ({ event }) =>
            api.events.system.ExtrinsicSuccess.is(event),
          );
          if (event) {
            resolve();
            unsub();
          }

          if (status.isFinalized) {
            reject(new Error("The event.ExtrinsicSuccess is not found"));
            unsub();
          }
        },
      );
    } catch (error) {
      reject(error);
    }
  });
};

export function defer<T>() {
  const deferred = {} as {
    resolve: (value: any) => void;
    reject: (reason: any) => void;
    promise: Promise<T>;
  };
  deferred.promise = new Promise((resolve, reject) => {
    deferred.resolve = resolve;
    deferred.reject = reject;
  });
  return deferred;
}

export const sendTransaction = async (tx: Promise<Extrinsic>) => {
  const signed = await tx;
  const deferred = defer<Codec[]>();
  await signed.send((status) => {
    if (status.isCompleted) {
      deferred.resolve(status.events);
    }
    if (status.isError) {
      deferred.reject(status.status);
    }
  });

  return {
    events: deferred.promise,
  };
};
