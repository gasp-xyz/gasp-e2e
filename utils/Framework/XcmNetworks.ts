import dotenv from "dotenv";
import { setupContext, SetupOption } from "./XcmHelper";

dotenv.config();

const endpoints = {
  kusama: "wss://kusama-rpc.polkadot.io",
  statemine: "wss://ksm-rpc.stakeworld.io/assethub",
  imbue: "wss://kusama.imbuenetwork.com",
  karura: "wss://karura-rpc-0.aca-api.network",
  mangata: "wss://kusama-archive.mangata.online",
  bifrost: "wss://bifrost-parachain.api.onfinality.io/public-ws",
  turing: "wss://rpc.turing.oak.tech",
  moonriver: "wss://moonriver-rpc.dwellir.com",
};

const toNumber = (value: string | undefined): number | undefined => {
  if (value === undefined) {
    return undefined;
  }

  return Number(value);
};

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  kusama: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.KUSAMA_WASM,
      blockNumber: toNumber(process.env.KUSAMA_BLOCK_NUMBER),
      endpoint: process.env.KUSAMA_ENDPOINT ?? endpoints.kusama,
      db: process.env.DB_PATH,
      ...options,
    }),
  imbue: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.IMBUE_WASM,
      blockNumber: toNumber(process.env.IMBUE_BLOCK_NUMBER),
      endpoint: process.env.IMBUE_ENDPOINT ?? endpoints.imbue,
      db: process.env.DB_PATH,
      ...options,
    }),
  turing: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.TURING_WASM,
      blockNumber: toNumber(process.env.TURING_BLOCK_NUMBER),
      endpoint: process.env.TURING_ENDPOINT ?? endpoints.turing,
      db: process.env.DB_PATH,
      ...options,
    }),
  statemine: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.STATEMINE_WASM,
      blockNumber: toNumber(process.env.STATEMINE_BLOCK_NUMBER),
      endpoint: process.env.STATEMINE_ENDPOINT ?? endpoints.statemine,
      db: process.env.DB_PATH,
      ...options,
    }),
  karura: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.KARURA_WASM,
      blockNumber: toNumber(process.env.KARURA_BLOCK_NUMBER),
      endpoint: process.env.KARURA_ENDPOINT ?? endpoints.karura,
      db: process.env.DB_PATH,
      ...options,
    }),
  biforst: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.BIFROST_WASM,
      blockNumber: toNumber(process.env.BIFROST_BLOCK_NUMBER),
      endpoint: process.env.BIFROST_ENDPOINT ?? endpoints.bifrost,
      db: process.env.DB_PATH,
      ...options,
    }),
  moonriver: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.MOONRIVER_WASM,
      blockNumber: toNumber(process.env.MOONRIVER_BLOCK_NUMBER),
      endpoint: process.env.MOONRIVER_ENDPOINT ?? endpoints.moonriver,
      db: process.env.DB_PATH,
      ...options,
    }),
  mangata: (options?: Partial<SetupOption>) =>
    setupContext({
      wasmOverride: process.env.MANGATA_WASM,
      blockNumber: toNumber(process.env.MANGATA_BLOCK_NUMBER),
      endpoint: process.env.MANGATA_ENDPOINT ?? endpoints.mangata,
      db: process.env.DB_PATH,
      types: {
        ShufflingSeed: {
          seed: "H256",
          proof: "H512",
        },
        Header: {
          parentHash: "Hash",
          number: "Compact<BlockNumber>",
          stateRoot: "Hash",
          extrinsicsRoot: "Hash",
          digest: "Digest",
          seed: "ShufflingSeed",
          count: "BlockNumber",
        },
      },
      ...options,
    }),
};
