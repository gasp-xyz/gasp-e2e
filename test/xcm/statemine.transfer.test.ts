import { connectParachains } from "@acala-network/chopsticks";
import { KeyringPair } from "@polkadot/keyring/types";
import { BN_THOUSAND } from "@polkadot/util";
import { AssetId } from "../../utils/ChainSpecs";
import { ApiContext } from "../../utils/Framework/XcmHelper";
import XcmNetworks from "../../utils/Framework/XcmNetworks";
import { devTestingPairs } from "../../utils/setup";
import { sendTransaction } from "../../utils/sign";
import {
  expectEvent,
  expectExtrinsicSuccess,
  expectJson,
  matchEvents,
  matchSystemEvents,
} from "../../utils/validators";

/**
 * @group xcm
 * @group proxied
 */
describe("XCM tests for Mangata <-> Statemine", () => {
  let statemine: ApiContext;
  let mangata: ApiContext;
  let alice: KeyringPair;

  beforeAll(async () => {
    statemine = await XcmNetworks.statemine();
    mangata = await XcmNetworks.mangata();
    await connectParachains([statemine.chain, mangata.chain]);
    alice = devTestingPairs().alice;
  });

  afterAll(async () => {
    await statemine.teardown();
    await mangata.teardown();
  });

  beforeEach(async () => {
    await mangata.dev.setStorage({
      Tokens: {
        Accounts: [
          [[alice.address, { token: 30 }], { free: 1000e6 }],
          [
            [alice.address, { token: 0 }],
            { free: AssetId.Mgx.unit.mul(BN_THOUSAND).toString() },
          ],
        ],
      },
      Sudo: {
        Key: alice.address,
      },
    });
    await statemine.dev.setStorage({
      System: {
        Account: [[[alice.address], { providers: 1, data: { free: 10e12 } }]],
      },
      Assets: {
        Account: [[[1984, alice.address], { balance: 1000e6 }]],
      },
    });
  });

  it("mangata transfer assets to statemine", async () => {
    const tx = await sendTransaction(
      mangata.api.tx.xTokens
        .transfer(
          30,
          10e6,
          {
            V1: {
              parents: 1,
              interior: {
                X2: [
                  { Parachain: 1000 },
                  {
                    AccountId32: {
                      network: "Any",
                      id: alice.addressRaw,
                    },
                  },
                ],
              },
            },
          },
          "Unlimited"
        )
        .signAsync(alice)
    );

    await mangata.chain.newBlock();

    expectExtrinsicSuccess(await tx.events);
    expectEvent(await tx.events, {
      event: expect.objectContaining({
        section: "xTokens",
        method: "TransferredMultiAssets",
      }),
    });

    await statemine.chain.newBlock();
    expectJson(
      await mangata.api.query.tokens.accounts(alice.address, 30)
    ).toMatchSnapshot();

    expect(
      await statemine.api.query.assets.account(1984, alice.address)
    ).toMatchSnapshot();

    await matchSystemEvents(mangata, "xcmpQueue", "Success");
  });

  it("statemine transfer assets to mangata", async () => {
    const tx = await sendTransaction(
      statemine.api.tx.polkadotXcm
        .limitedReserveTransferAssets(
          {
            V3: {
              parents: 1,
              interior: {
                X1: { Parachain: 2110 },
              },
            },
          },
          {
            V3: {
              parents: 0,
              interior: {
                X1: {
                  AccountId32: {
                    id: alice.addressRaw,
                  },
                },
              },
            },
          },
          {
            V3: [
              {
                id: {
                  Concrete: {
                    parents: 0,
                    interior: {
                      X2: [{ PalletInstance: 50 }, { GeneralIndex: 1984 }],
                    },
                  },
                },
                fun: { Fungible: 10e6 },
              },
            ],
          },
          0,
          "Unlimited"
        )
        .signAsync(alice, { nonce: 0 })
    );

    await statemine.chain.newBlock();

    await matchEvents(tx.events, "polkadotXcm");

    expect(
      await statemine.api.query.assets.account(1984, alice.address)
    ).toMatchSnapshot();

    await mangata.chain.newBlock();

    expectJson(
      await mangata.api.query.tokens.accounts(alice.address, 30)
    ).toMatchSnapshot();

    await matchSystemEvents(mangata, "xcmpQueue", "Success");
  });
});
