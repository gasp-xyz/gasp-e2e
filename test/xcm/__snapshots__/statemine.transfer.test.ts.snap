// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XCM tests for Mangata <-> Statemine mangata transfer assets to statemine 1`] = `
{
  "free": 990000000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`XCM tests for Mangata <-> Statemine mangata transfer assets to statemine 2`] = `
{
  "balance": 1009996407,
  "extra": null,
  "isFrozen": false,
  "reason": {
    "consumer": null,
  },
}
`;

exports[`XCM tests for Mangata <-> Statemine mangata transfer assets to statemine: system events 1`] = `
[
  {
    "data": {
      "messageHash": "(redacted)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;

exports[`XCM tests for Mangata <-> Statemine statemine transfer assets to mangata 1`] = `
{
  "balance": 990000000,
  "extra": null,
  "isFrozen": false,
  "reason": {
    "consumer": null,
  },
}
`;

exports[`XCM tests for Mangata <-> Statemine statemine transfer assets to mangata 2`] = `
{
  "free": 1009991936,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`XCM tests for Mangata <-> Statemine statemine transfer assets to mangata: events 1`] = `
[
  {
    "data": [
      {
        "Complete": {
          "proofSize": "(redacted)",
          "refTime": "(redacted)",
        },
      },
    ],
    "method": "Attempted",
    "section": "polkadotXcm",
  },
]
`;

exports[`XCM tests for Mangata <-> Statemine statemine transfer assets to mangata: system events 1`] = `
[
  {
    "data": {
      "messageHash": "(redacted)",
      "weight": {
        "proofSize": "(redacted)",
        "refTime": "(redacted)",
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;
